README.md
# F is for format & WHAT THE diff (& Friends!)

A JavaScript Library that extends JavaScript's Date object to provide all the
functionality that is missing from the standard implementation. With **F is for
Format** you can easily output a date in almost any format. **WHAT THE diff**
gives you the ability to calculate the exact difference between *any* two dates
and output the difference in human readable text. Also included are several
other useful Date methods including **Date.isLeapYear** and
**Date.getDaysInMonth**.


## F is for Format

### Date.f(format)

Outputs a JavaScript Date Object in virtually any textual format.

**format**\
Type: String\
A string specifying how the dat should be formatted

The following characters are recognized in the format parameter string:
|Character|Description|Example returned values|
|--- |--- |--- |
|Day|---|---|
|d|Day of the month, 2 digits with leading zeros|01 to 31|
|D|A textual representation of a day, three letters|Mon through Sun|
|j|Day of the month without leading zeros|1 to 31|
|l (lowercase 'L')|A full textual representation of the day of the week|Sunday through Saturday|
|ll (lowercase 'LL')|Same as l, but with contextual labels for today,tomorrow, and yesterday|Sunday through Saturday, Today, Tomorrow, Yesterday|
|N|ISO-8601 numeric representation of the day of the week|1 (for Monday) through 7 (for Sunday)|
|S|English ordinal suffix for the day of the month, 2 characters|st, nd, rd or th.  Works well with j|
|w|Numeric representation of the day of the week|0 (for Sunday) through 6 (for Saturday)|
|z|The day of the year (starting from 0)|0 through 365|
|Week|---|---|
|W|ISO-8601 week number of year, weeks starting on Monday|Example: 42 (the 42nd week in the year)|
|Month|---|---|
|F|A full textual representation of a month, such as January or March|January through December|
|m|Numeric representation of a month, with leading zeros|01 through 12|
|M|A short textual representation of a month, three letters|Jan through Dec|
|n|Numeric representation of a month, without leading zeros|1 through 12|
|t|Number of days in the given month|28 through 31|
|Year|---|---|
|L|Whether it's a leap year|1 if it is a leap year, 0 otherwise.|
|o|ISO-8601 week-numbering year. This has the same value as Y, except that if the ISO week number (W) belongs to the previous or next year, that year is used instead.|Examples: 1999 or 2003|
|Y|A full numeric representation of a year, 4 digits|Examples: 1999 or 2003|
|y|A two digit representation of a year|Examples: 99 or 03|
|Time|---|---|
|a|Lowercase Ante meridiem and Post meridiem|am or pm|
|A|Uppercase Ante meridiem and Post meridiem|AM or PM|
|B|Swatch Internet time|000 through 999|
|g|12-hour format of an hour without leading zeros|1 through 12|
|G|24-hour format of an hour without leading zeros|0 through 23|
|h|12-hour format of an hour with leading zeros|01 through 12|
|H|24-hour format of an hour with leading zeros|00 through 23|
|i|Minutes with leading zeros|00 to 59|
|ii|Minutes without leading zeros|0 to 59|
|s|Seconds with leading zeros|00 through 59|
|ss|Seconds without leading zeros|0 through 59|
|v|Milliseconds|Example: 654|
|Timezone|---|---|
|e|Timezone identifier|Examples: Pacific Daylight Time, Pacific Standard Time|
|I (capital i)|Whether or not the date is in daylight saving time|1 if Daylight Saving Time, 0 otherwise.|
|O|Difference to Greenwich time (GMT) without colon between hours and minutes|Example: +0200|
|P|Difference to Greenwich time (GMT) with colon between hours and minutes|Example: +02:00|
|p|The same as P, but returns Z instead of +00:00|Example: +02:00|
|Z|Timezone offset in seconds. The offset for timezones west of UTC is always negative, and for those east of UTC is always positive.|-43200 through 50400|
|Full Date/Time|---|---|
|c|ISO 8601 date|2004-02-12T15:19:21+00:00|
|r|» RFC 2822 formatted date|Example: Thu, 21 Dec 2000 16:01:07 +0200|
|U|Seconds since the Unix Epoch (January 1 1970 00:00:00 GMT)|1627616755444|

Example:
```javascript
var myDate = new Date();
var myFormattedDate = myDate.f("l, F j, y");
var myFormattedDate2 = myDate.f("m/d/Y H:i:s");
alert("Today is: " + myFormattedDate);
alert("Or would you prefer something more specific: " + myFormattedDate2);
```



## WHAT THE diff??

### Date.diff([date], [settings])

Calculates the exact difference between any two dates, broken down into customizable increments.

**date**\
Type: Date object, valid date string, or Unix epoch timestamp\
*Optional*  The datetime to compare against.  If not specified the current time will be used.

**settings**\
Type: PlainObject or valid unit string\
*Optional*  A set of key/value pairs that define how the difference should be displayed.

- **abbr**  *(dafault: FALSE)*\
  Type: Boolean\
  Use abbreviated time unit names
- **divider**  *(default: ", ")*\
  Type: String\
  Specifies the divider to place between units of time
- **labels**\
  Type: PlainObject, Boolean\
  A set of key value pairs that can be used to override the default unit labels.  Pass FALSE to not display any labels.
- **lc**  *(default: FALSE)*\
  Type: Boolean\
  When set to TRUE, all unit labels will be displayed in lowercase
- **same**  *(default: "Same")*\
  Type: String\
  A string specifying the label to use when the two dates are identical
- **units**  *(default: "*")*\
  Type: String\
  A string specifying which units of time to use to display the difference
- **zeros** *(default: FALSE)*\
  Type: Boolean\
  When set to TRUE, leading zeros will be displayed for all applicable time units

The following characters are used to specify the units of time to display:
|Character|Unit of Time|
|---|--- |
|T|Millenia|
|C|Centuries|
|D|Decades|
|Y|Years|
|M|Months|
|W|Weeks|
|d|Days|
|H|Hours|
|m|Minutes|
|S|Seconds|
|N|Miliseconds|
|*|All (default)|


Example:
```javascript
var oldDate = new Date("03/22/2009 23:23:12");
var curDate = new Date();
var theDiff = curDate.diff(oldDate);
//is the same as
var theDiff2 = curDate.diff(oldDate, "TCDYMWdHmSN");
alert("And the difference is: " + theDiff);

var theDiff3 = curDate.diff(oldDate, {
   units : "dS",
   abbr  : true,
   divider : " - "
});
alert("The diff in just days and seconds is: " + theDiff3);
```


## (& Friends!)

### Date.getDaylightSavingsDays()

Returns an array of two date objects for the two daylight savings change days within the year of the provided date

Example:
```javascript
var thisYear = new Date();
alert(
   "On " + thisYear.getDaylightSavingsDays[0].f("DD, MMM d") + " spring forward.\n" 
   + "On " + thisYear.getDaylightSavingsDays[1].f("DD, MMM d") + " fall back."
);
```

### Date.getDaysInMonth()

Returns the number of days in the selected month

Example:
```javascript
var thisMonth = new Date();
var daysThisMonth = thisMonth.getDaysInMonth();
alert("There are " + daysThisMonth + " days in this month.");
```

### Date.isDaylightSavings()

Returns TRUE if the daylight savings is in effect for the date provided, otherwise FALSE

Example:
```javascript
var tomorrow = new Date();
tomorrow.setDate(+1);
alert("Daylight Savings is " + (tomorrow.isDaylightSavings() ? "" : "not ") + "currently in effect!");
```

### Date.isDaylightSavingsDay()

Returns TRUE if the provided date is a daylight savings change day, otherwise FALSE

Example:
```javascript
var tomorrow = new Date();
tomorrow.setDate(+1);
if(tomorrow.isDaylightSavingsDay()) alert("Don't forget to change the clock tonight!");
```

### Date.isLeapYear()

Returns TRUE if the selected year is a leap year, otherwise FALSE

Example:
```javascript
var thisYear = new Date();
if(thisYear.isLeapYear()) alert("It's a leap year!");
else {
   alert("Maybe next year..."); 
   thisYear.setYear(+1); 
   if(thisYear.isLeapYear()) alert("Yes!!");                                                   
   else alert("Darn.");                           
}
```

### Date.when()

 Returns a historical timestamp indicating how long ago the provided date was, like the timestamps on social media posts  (Note: Future dates will be labeled "Now")
 
 Example:
 ```javascript
var now = new Date();
alert("This date is " + now.when());
now.setHours(-2);
alert("But this date is from " + now.when());
```



CHANGELOG
# "f" is for Format & What the diff?? Changelog [Updated 2021-01-28]

[2021-07-27] Version 1.2.1
   - Code cleanup
   - NPM build

[2021-01-28] Version 1.2
   - Various improvements
   
[2009-07-20] Version 0.5.0
   - Various improvements
   
[2009-03-20] Version 0.3.0
   - Fixed error caused by using the reserved "short" parameter
   - Improved variable loading
   
[2009-03-03] Version 0.1.0
   - Initial release

date.f.js
/*!
 * F is for Format, WHAT THE diff??, & Friends v1.2
 *
 * Kopimi 2021 Joshua Faulkenberry
 * Unlicensed under The Unlicense
 * http://unlicense.org/
 */
 
/************** F is for Format ***************
 * Outputs a JavaScript Date Object in various
 * customizable formats
 **********************************************
 */
 
window.Date.prototype.f = function(format) {
   if(this === "" || this == "NaN") {
      return "Invalid Date";
   }
   if(format == "@") {
      return this.getTime();
   }
   else if(format == "REL") {
      var diff = (((new Date()).getTime() - this.getTime()) / 1000), day_diff = Math.floor(diff / 86400);
      return day_diff === 0 && (
          diff > -60 && "right now" ||
         diff > -120 && "1 minute from now" ||
         diff > -3600 && -(Math.floor(diff / 60)) + " minutes from now" ||
         diff > -7200 && "1 hour ago" ||
         diff > -86400 && -(Math.floor(diff / 3600)) + " hours from now" ||

         diff < 60 && "just now" ||
         diff < 120 && "1 minute ago" ||
         diff < 3600 && Math.floor( diff / 60 ) + " minutes ago" ||
         diff < 7200 && "1 hour ago" ||
         diff < 86400 && Math.floor( diff / 3600 ) + " hours ago") ||

         day_diff === 0 && "Tomorrow" ||
         day_diff > -7 && -(day_diff) + " days from now" ||
         -(Math.ceil( day_diff / 7 )) == 1 && "1 week from now" ||
         day_diff > -78 && -(Math.ceil( day_diff / 7 )) + " weeks from now" ||
         day_diff > -730 && -(Math.ceil( day_diff / 30 )) + " months from now" ||
         day_diff <= -730 && -(Math.ceil( day_diff / 365 )) + " years from now" ||

         day_diff == 1 && "Yesterday" ||
         day_diff < 7 && day_diff + " days ago" ||
         (Math.ceil( day_diff / 7 )) == 1 && "1 week ago" ||
         day_diff < 78 && Math.ceil( day_diff / 7 ) + " weeks ago" ||
         day_diff < 730 && Math.ceil( day_diff / 30 ) + " months ago" ||
         Math.ceil( day_diff / 365 ) + " years ago";
   }
   var MONTH_NAMES=['January','February','March','April','May','June','July','August','September','October','November','December'],
       DAY_NAMES=['Sunday','Monday','Tuesday','Wednesday','Thursday','Friday','Saturday'],
       date = this,
       result="",
       i_format=0,
       c="",
       token="",
       y=date.getYear()+"",
       M=date.getMonth()+1,
       d=date.getDate(),
       E=date.getDay(),
       D=date.getDay(),
       H=date.getHours(),
       m=date.getMinutes(),
       s=date.getSeconds(),
       n=date.getMilliseconds(),
       yyyy,yy,MMM,MM,dd,hh,h,mm,ss,nn,ampm,HH,KK,K,kk,k,
       LZ = function(x) {
           return(x<0||x>9?"":"0")+x;
       },
       midnight = (new Date());
       if(!MONTH_NAMES[M-1]) return "Invalid Date";
   format=format+"";
   midnight.setHours(0);
   midnight.setMinutes(0);
   midnight.setSeconds(0);
   midnight.setMilliseconds(0);
   if (y.length < 4) {
      y=""+(y-0+1900);
   }
   var value = {
      y: ""+y,
      yyyy: y,
      yy: (y+"").substr(2,4),
      M: M,
      MM: LZ(M),
      MMM: MONTH_NAMES[M-1],
      NNN: MONTH_NAMES[M-1].substr(0,3),
      N: MONTH_NAMES[M-1].substr(0,1),
      d: d,
      dd: LZ(d),
      e: DAY_NAMES[E].substr(0,1),
      ee: DAY_NAMES[E].substr(0,2),
      E: DAY_NAMES[E].substr(0,3),
      EE: DAY_NAMES[E],
      D: (date.getTime() >= midnight.getTime() && date.getTime() < (midnight.getTime() + (1000*60*60*24))) ? "Today" : DAY_NAMES[D].substr(0,3),
      DD: (date.getTime() >= midnight.getTime() && date.getTime() < (midnight.getTime() + (1000*60*60*24))) ? "Today" : DAY_NAMES[D],
      H: H,
      HH: LZ(H),
      h: (H===0) ? 12 : (H>12) ? H-12 : H,
      hh: LZ((H===0) ? 12 : (H>12) ? H-12 : H),
      k: H+1,
      kk: LZ(H+1),
      K: (H>11) ? H-12 : H,
      KK: LZ((H>11) ? H-12 : H),
      m: m,
      mm: LZ(m),
      s: s,
      ss: LZ(s),
      n: n,
      a: (H>11) ? "PM" : "AM",
   };
   while (i_format < format.length) {
      c=format.charAt(i_format);
      token="";
      while ((format.charAt(i_format)==c) && (i_format < format.length)) {
         token += format.charAt(i_format++);
      }
      if (value[token] != null) { result=result + value[token]; }
      else { result=result + token; }
   }
   return result;
};

/************ WHAT'S THE diff?? *************
 * Calculates the exact difference between
 * any two dates and outputs the results in
 * a customizable incremental breakdown
 ********************************************
 */
window.Date.prototype.diff = function(date, options) {
   if(typeof(date) == "undefined") date = new Date();
   else if(typeof(date) ==  "string" || typeof(date) ==  "number") {
      if((new Date(date)) != "Invalid Date" && (new Date(date)) != "NaN") {
         date = new Date(date);
      }
      else {
         console.log("Invalid diff comparison date: " + date);
         return false;
      }
   }
   else if(typeof(date) ==  "object" && !date.getTime) {
      options = date;
      date = new Date();
   }
   if(typeof(options) ==  "string") {
      options = {
         breakdown: options
      };
   }
   if(!options) options = {};
   options = {
      breakdown: options.breakdown || "TCDYMWdHmSN",
      divider:   options.divider || ", ",
      abbr:      options.abbr || false,
      same:      options.same || "Same",
      zeros:     options.zeros || false,
      labels:    options.hasOwnProperty('labels') ? options.labels : {
         T : ["Millennium","Millennia","ml"],
         C : ["Century","Centuries","c"],
         D : ["Decade","Decades","d"],
         Y : ["Year","Years","yr"],
         M : ["Month","Months","mo"],
         W : ["Week","Weeks","wk"],
         d : ["Day","Days","d"],
         H : ["Hour","Hours","hr"],
         m : ["Minute","Minutes","m"],
         S : ["Second","Seconds","s"],
         N : ["Millisecond","Milliseconds","ms"]
      }
   };
   var min   = date <= this && date || date > this && this,
       max   = date > this && date || date <= this && this,
       diff  = (max.getTime() - min.getTime()),
       tl    = {
          T : 1000*60*60*24*365*100*10,
          C : 1000*60*60*24*365*100,
          D : 1000*60*60*24*365*10,
          Y : 1000*60*60*24*365,
          M : 1000*60*60*24*30,
          W : 1000*60*60*24*7,
          d : 1000*60*60*24,
          H : 1000*60*60,
          m : 1000*60,
          S : 1000,
          N : 1
       },
       result = [];

   for(var i = 0; i < options.breakdown.length; i++) {
      var res = Math.floor(diff / tl[options.breakdown.charAt(i)]);
      if(res || options.zeros) {
         if(options.zeros && options.breakdown.charAt(i) == "N") {
            if(res < 10) res = "00" + res;
            else if(res < 100) res = "0" + res;
         }
         result.push(res + (options.labels ? " " + options.labels[options.breakdown.charAt(i)][options.abbr ? 2 : res == 1 ? 0 : 1] : ""));
         diff = diff % tl[options.breakdown.charAt(i)];
      }
   };

   diff = result.join(options.divider);
   if(diff === "") {
      diff = options.same;
   }
   if(options.lc) {
      diff = diff.toLowerCase();
   }
   return diff;
};

/*************** Date.when() ****************
 * Returns a fancy timestamp indicating how
 * long ago a date was like you see on
 * Facebook or Twitter
 ********************************************
 */
window.Date.prototype.when = function() {
   var
      now       = new Date(),
      ts        = this.f("MMM d, yyyy h:mm a"),
      nowSec    = now.getTime() / 1000,
      tsSec     = this.getTime() / 1000;

   if((nowSec - tsSec) < 60) ts = "Now";
   else if((nowSec - tsSec) / 60 < 60) ts = this.diff(now, "m");
   else if(((nowSec - tsSec) / 60) / 60 < 12) ts = this.diff(now, "H");
   else if(now.f("yyyyMMdd") == this.f("yyyyMMdd")) ts = "Today" + this.f("h:mm a");
   else if(tsSec > (new Date(new Date(now.f("MM/dd/yyyy 00:00:00")).setDate(now.getDate()-1))).getTime() / 1000) ts = "Yesterday " + this.f("h:mm a");
   else if(tsSec > (new Date(new Date(now.f("MM/dd/yyyy 00:00:00")).setDate(now.getDate() - now.getDay()))).getTime() / 1000) ts = this.f("EE h:mm a");
   else if(tsSec > (new Date(new Date().getFullYear(), 0, 1)).getTime() / 1000) ts = this.f("MMM d h:mm a");
   return ts;
};

/********* Date.getDaysInMonth() *************
 * Returns the number of days in the current
 * month
 *********************************************
 */
window.Date.prototype.getDaysInMonth = function() {
   return [31,28,31,30,31,30,31,31,30,31,30,31][this.getMonth()];
};

/************* Date.isLeapYear() ***************
 * Returns true if the current year is a leap
 * year
 ***********************************************
 */
window.Date.prototype.isLeapYear = function() {
   return (new Date(this.getFullYear(),1,29)).getDate() == 29;
};

/******* Date.getDayLightSavingsDays() *********
 * Returns an array containing date objects for
 * the two daylight savings change days within
 * the current year
 ***********************************************
 */
window.Date.prototype.getDayLightSavingsDays = function() {
   var result = [];
   var day1 = new Date("03/07/"+this.getFullYear());
   var day2 = new Date("03/06/"+this.getFullYear());
   while(day1.getMonth() < 3 || (day1.getMonth() == 3  && day1.getDate() < 16)) {
      if((day1.getTime() - day2.getTime())/1000/60/60 != 24) {
         result[result.length] = new Date(day2.getTime());
      }
      day1.setDate(day1.getDate()+1);
      day2.setDate(day2.getDate()+1);
   }
   day1 = new Date("10/31/"+this.getFullYear());
   day2 = new Date("10/30/"+this.getFullYear());
   while(day1.getMonth() < 11 || (day1.getMonth() == 10 && day1.getDate() < 9)) {
      if((day1.getTime() - day2.getTime())/1000/60/60 != 24) {
         result[result.length] = new Date(day2.getTime());
      }
      day1.setDate(day1.getDate()+1);
      day2.setDate(day2.getDate()+1);
   }
   return result;
};

/******** Date.isDayLightSavingsDay() **********
 * Returns true if the current day is a
 * daylight savings change day
 ***********************************************
 */
window.Date.prototype.isDayLightSavingsDay = function() {
   var comp = new Date(this.getTime());
   comp.setDate(comp.getDate()+1);
   return (comp.getTime() - this.getTime())/1000/60/60 != 24;
};
